src/mlpack/core/optimizers/pso/pso.hpp

```diff
@@ -0,0 +1,188 @@
+/**
+ * @file pso.hpp
+ * @author Adeel Ahmad
+ *
+ * Definition of the Particle Swarm Optimizer as proposed
+ * by J. Kennedy et al. in "Particle swarm optimization".
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PSO_PSO_HPP
+#define MLPACK_CORE_OPTIMIZERS_PSO_PSO_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "inertia_weight.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * PSO - Particle swarm optimization is a method based on  the social behavior
+ * of bird flocks when moving from one place to another and was proposed
+ * mainly to solve numerical optimization problems. PSO with intertia weight
+ * variant is implemented in the code below.
+ *
+ * For more information, please refer to:
+ *
+ * @code
+ * @article{Kennedy1995
+ *   author    = {J. Kennedy and R. Eberhart},
+ *   title     = {Particle swarm optimization},
+ *   volume    = {4},
+ *   number    = {},
+ *   year      = {1995},
+ *   pages     = {1942-1948 vol.4},
+ *   publisher = {},
+ * }
+ * @endcode
+ *
+ * For PSO to work, the class must implement the following function:
+ *
+ *   size_t NumFunctions();
+ *   double Evaluate(const arma::mat& coordinates, const size_t i);
+ *
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function is
+ * being evaluated.  So, for the case of a data-dependent function, such as NCA
+ * (see mlpack::nca::NCA), NumFunctions() should return the number of points in
+ * the dataset, and Evaluate(coordinates, 0) will evaluate the objective function
+ * on the first point in the dataset (presumably, the dataset is held internally
+ * in the DecomposableFunctionType).
+ *
+ * @tparam SelectionPolicy The velocty / position update strategy used for the
+ * evaluation step.
+ */
+template<typename VelocityVectorType = InertiaWeight>
+class PSO
```

Comment:
This is just an idea but we could name the class ``PSOType`` and define an alias called ``PSO``, that way we could use ``PSO`` instead of ``PSO<>`` later on, which makes more sense if we introduce another velocity type.